clients/unshaded/src/main/java/tachyon/client/lineage/LineageMasterClient.java

```diff
@@ -0,0 +1,159 @@
+ * the License.
+ */
+
+package tachyon.client.lineage;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+
+import tachyon.Constants;
+import tachyon.MasterClientBase;
+import tachyon.TachyonURI;
+import tachyon.conf.TachyonConf;
+import tachyon.job.CommandLineJob;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.LineageDeletionException;
+import tachyon.thrift.LineageDoesNotExistException;
+import tachyon.thrift.LineageInfo;
+import tachyon.thrift.LineageMasterService;
+
+/**
+ * A wrapper for the thrift client to interact with the lineage master, used by tachyon clients.
+ *
+ * Since thrift clients are not thread safe, this class is a wrapper to provide thread safety, and
+ * to provide retries.
+ */
+public final class LineageMasterClient extends MasterClientBase {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private LineageMasterService.Client mClient = null;
+
+  /**
+   * Creates a new lineage master client.
+   *
+   * @param masterAddress the master address
+   * @param executorService the executor service
+   * @param tachyonConf the Tachyon configuration
+   */
+  public LineageMasterClient(InetSocketAddress masterAddress, ExecutorService executorService,
+      TachyonConf tachyonConf) {
+    super(masterAddress, executorService, tachyonConf);
+  }
+
+  @Override
+  protected String getServiceName() {
+    return Constants.LINEAGE_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  protected void afterConnect() {
+    mClient = new LineageMasterService.Client(mProtocol);
+  }
+
+  public synchronized long createLineage(List<TachyonURI> inputFiles, List<TachyonURI> outputFiles,
+      CommandLineJob job) throws IOException, FileDoesNotExistException {
+    // prepare for RPC
+    List<String> inputFileStrings = Lists.newArrayList();
+    for (TachyonURI inputFile : inputFiles) {
+      inputFileStrings.add(inputFile.toString());
+    }
+    List<String> outputFileStrings = Lists.newArrayList();
+    for (TachyonURI outputFile : outputFiles) {
+      outputFileStrings.add(outputFile.toString());
+    }
+
+    int retry = 0;
+    while (!mClosed && (retry ++) <= RPC_MAX_NUM_RETRY) {
+      connect();
+      try {
+        return mClient.createLineage(inputFileStrings, outputFileStrings,
+            job.generateCommandLineJobInfo());
+      } catch (FileDoesNotExistException e) {
+        throw e;
+      } catch (TException e) {
+        LOG.error(e.getMessage(), e);
+        mConnected = false;
+      }
+    }
+    throw new IOException("Failed after " + retry + " retries.");
+  }
+
+  public synchronized boolean deleteLineage(long lineageId, boolean cascade)
```

Comment:
hmm cascade delete is a well-known concept in database that deletes a DAG node. I use it here to resemble it. What do you think?